r55/src/exec.rs

```diff
@@ -182,33 +241,23 @@ fn execute_riscv(
                 let t0: u64 = emu.cpu.xregs.read(5);
 
                 let Ok(syscall) = Syscall::try_from(t0 as u8) else {
-                    println!("Unhandled syscall: {:?}", t0);
-                    return return_revert(interpreter, evm_gas);
+                    warn!("Unhandled syscall: {:?}", t0);
+                    return return_revert(interpreter, interpreter.gas.spent());
                 };
+                debug!("[Syscall::{} - {:#04x}]", syscall, t0);
 
                 match syscall {
                     Syscall::Return => {
                         let ret_offset: u64 = emu.cpu.xregs.read(10);
                         let ret_size: u64 = emu.cpu.xregs.read(11);
 
                         let r55_gas = r55_gas_used(&emu.cpu.inst_counter);
-                        let data_bytes = dram_slice(emu, ret_offset, ret_size)?;
+                        debug!("> Total R55 gas: {}", r55_gas);
 
-                        let total_cost = r55_gas + evm_gas;
-                        println!(
-                            "evm gas: {}, r55 gas: {}, total cost: {}",
-                            evm_gas, r55_gas, total_cost
-                        );
-                        let in_limit = interpreter.gas.record_cost(total_cost);
-                        if !in_limit {
-                            return Ok(InterpreterAction::Return {
-                                result: InterpreterResult {
-                                    result: InstructionResult::OutOfGas,
-                                    output: Bytes::new(),
-                                    gas: interpreter.gas,
-                                },
-                            });
-                        }
+                        // RETURN logs the gas of the whole risc-v instruction set
+                        syscall_gas!(interpreter, r55_gas);
```

Comment:
seems fine right?

Reply:
when testing i made sure that i got the exact same gas costs as the original implementation.
as mentioned above, this approach fixes the scenario where there are multiple calls (execution contexts)